### ISSUE TITLE
Is the formula scale invariant?

### ISSUE THREAD

**Issue by Yuancheng-Xu** [eth-sri/language-model-arithmetic#4]
Hello @JasperDekoninck ,

I wonder if the formula is scale invariant. For example, is `2 * M_1 + 1 * M_2` the same as `4 * M_1 + 2 * M_2`? From the paper, I suppose it is. However, I wonder if it is true for the implementation. Thank you in advance!

**Comment by JasperDekoninck** [comment]
The formula is indeed scale invariant, also in the implementation. 

The normalization has to be done to avoid sampling from a distribution with a very high entropy. The normalization can be explained as some kind of temperature discount: if there was no normalization, then sampling from the first formula at temperature 1 / (1+2) would be equivalent to sampling from the second formula with temperature 1/(4+2). Therefore, if you need to have the unnormalized version of your formula, you can simply sample with temperature 3 and temperature 6 from the respective formulas. Note that this would lead to "garbage" output, since this would behave similar to sampling at these high temperatures from "normal" language models.

**Comment by Yuancheng-Xu** [comment]
Got it. Thank you!

**Comment by Yuancheng-Xu** [comment]
Hi @JasperDekoninck I am reading your code in details and I really like it! I have the following questions and I appreciate any help!

## Question 1
Given two LLMs $\pi_1$ (M_1) and $\pi_2$ (M_2), Using model arithmetic formula `M_1 + 2 * M_2` (and setting temperature = 1 during generation), my understanding (is it correct?) is that the code base implements the LLM $\pi'$ which satisfies

$$\log \pi'(y|x) =\frac{1}{3} \log \pi_1(y|x) + \frac{2}{3}\log \pi_2(y|x) + c(x).$$

I am wondering how to implement the following LLM $\pi$, without the normalization, using the codebase.

$$\log \pi(y|x) = \log \pi_1(y|x) + 2 \log \pi_2(y|x) + c(x)$$

Is the codebase directly applicable? For example, if I still use `M_1 + 2 * M_2` but also use temperature = $\frac{1}{3}$ in `generate_text`, do I obtain $\pi$?

## Question 2

Another related question is: My understanding is that when implementing `M = M_1 + 2 * M_2`, your codebase 
1. Compute logits_1 and logits_2.
2. Apply log_softmax (with temperature=1) to them and obtain log_prob_1 and log_prob_2.
3. Compute logit_M = 1/3 * log_prob_1 + 2/3 * log_prob_2.
4. Apply log_softmax (with temperature=1) to logit_M and obtain log_prob_M.
5. Apply softmax (with temperature specified in `generate_text`) to log_prob_M to obtain the sampling distribution for next-token generation.

It seems that my understanding of the code is probably not correct, since it is not consistent with how a single LLM is implemented, where the temperature is applied to logit, not log_prob. I wonder what is the correct pipeline when implementing `M = M_1 + 2 * M_2`.

Thank you!!

**Comment by JasperDekoninck** [comment]
Of course :)

## Question 1
You are correct in your understanding that the code returns $\pi'$. As shown in our paper, this normalization is required to obtain the correct solution to the KL-minimization problem. If you want to obtain $\pi$, you should indeed run `generate_text` with `temperature=1/3` (in my previous reply I accidentally claimed the equivalent of having to set `temperature=3` in this case, but that was an error). Temperature is essentially an inverse multiplier of the log probabilities, so the results would be the renormalization of $3 \log \pi'$. Note that this is not exactly equal to your $\pi$ since the normalization $c(x)$ will be different between $\pi$ and $\pi'$, but this difference is always necessary to obtain a valid probability distribution.

Note that if you run it with `temperature=1/3` you will get the feeling as if you were interacting with a model with a very low temperature (not a lot of variety). The distribution will be a lot more peaked than the individual models (that is why the normalization is useful).

## Question 2
Your understanding of the code is correct. There are two reasons why our pipeline is the correct one. (1) it is the exact implementation of the closed solution to the KL-minimization problem in our paper and (2) they are equivalent! It takes some vector calculus to work this out, but applying the temperature on the logits or the logprobs is equivalent, as long as you renormalize after (which we do). Instead of giving you the proof (which is a bit cumbersome), you can run the following code for some random vector to show that this is indeed the case!

```python
import torch

logits = torch.rand(10)
temperature = 2
print(torch.log_softmax(logits / temperature, 0))
# -> tensor([-2.3891, -2.1712, -2.5899, -2.2133, -2.2057, -2.2683, -2.3315, -2.4400,  -2.1951, -2.2971])
print(torch.log_softmax(torch.log_softmax(logits, 0) / temperature, 0))
# -> tensor([-2.3891, -2.1712, -2.5899, -2.2133, -2.2057, -2.2683, -2.3315, -2.4400,  -2.1951, -2.2971])
```

**Comment by Yuancheng-Xu** [comment]
@JasperDekoninck  Thank you!

Yep indeed I just realize that, mathematically logprob - logit = a constant vector (one-vector times a constant). Therefore applying softmax using any temperature to both of them will have the same probability distribution. Thank you for pointing it out!
